pkg/gatewayroute/resource_manager.go

```diff
@@ -0,0 +1,357 @@
+	"github.com/aws/aws-app-mesh-controller-for-k8s/pkg/references"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/pkg/runtime"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/pkg/virtualgateway"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/pkg/virtualservice"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	appmeshsdk "github.com/aws/aws-sdk-go/service/appmesh"
+	"github.com/go-logr/logr"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ResourceManager is dedicated to manage AppMesh GatewayRoute resources for k8s GatewayRoute CRs.
+type ResourceManager interface {
+	// Reconcile will create/update AppMesh GatewayRoute to match gr.spec, and update gr.status
+	Reconcile(ctx context.Context, gr *appmesh.GatewayRoute) error
+
+	// Cleanup will delete AppMesh GatewayRoute created for gr.
+	Cleanup(ctx context.Context, gr *appmesh.GatewayRoute) error
+}
+
+func NewDefaultResourceManager(
+	k8sClient client.Client,
+	appMeshSDK services.AppMesh,
+	referencesResolver references.Resolver,
+	accountID string,
+	log logr.Logger) ResourceManager {
+
+	return &defaultResourceManager{
+		k8sClient:          k8sClient,
+		appMeshSDK:         appMeshSDK,
+		referencesResolver: referencesResolver,
+		accountID:          accountID,
+		log:                log,
+	}
+}
+
+// defaultResourceManager implements ResourceManager
+type defaultResourceManager struct {
+	k8sClient          client.Client
+	appMeshSDK         services.AppMesh
+	referencesResolver references.Resolver
+	accountID          string
+	log                logr.Logger
+}
+
+func (m *defaultResourceManager) Reconcile(ctx context.Context, gr *appmesh.GatewayRoute) error {
+	ms, err := m.findMeshDependency(ctx, gr)
+	if err != nil {
+		return err
+	}
+	if err := m.validateMeshDependency(ctx, ms); err != nil {
+		return err
+	}
+	vg, err := m.findVirtualGatewayDependency(ctx, gr)
+	if err != nil {
+		return err
+	}
+	if err := m.validateVirtualGatewayDependency(ctx, ms, vg); err != nil {
+		return err
+	}
+	vsByKey, err := m.findVirtualServiceDependencies(ctx, gr)
+	if err != nil {
+		return err
+	}
+	if err := m.validateVirtualServiceDependencies(ctx, ms, vsByKey); err != nil {
+		return err
+	}
+
+	sdkGR, err := m.findSDKGatewayRoute(ctx, ms, vg, gr)
+	if err != nil {
+		return err
+	}
+	if sdkGR == nil {
+		sdkGR, err = m.createSDKGatewayRoute(ctx, ms, vg, gr, vsByKey)
+		if err != nil {
+			return err
+		}
+	} else {
+		sdkGR, err = m.updateSDKGatewayRoute(ctx, sdkGR, ms, vg, gr, vsByKey)
+		if err != nil {
+			return err
+		}
+	}
+
+	return m.updateCRDGatewayRoute(ctx, gr, sdkGR)
+}
+
+func (m *defaultResourceManager) Cleanup(ctx context.Context, gr *appmesh.GatewayRoute) error {
+	ms, err := m.findMeshDependency(ctx, gr)
+	if err != nil {
+		return err
+	}
```

Comment:
So, if there is no virtualgateway associated with these set of gateway routes then can't we go ahead and delete it? Is there a reason we block the delete for a mesh being active?

Reply:
No, my question is why do we need to check if there is an active mesh in the namespace before deleting gateway routes? Shouldn't we just check for virtual gateway? I can just choose to remove VG and GR and should be able to continue to use the mesh like how we do currently, right?